src/general/world_effects/patch_events/GlobalGlaciationEvent.cs

```diff
@@ -0,0 +1,258 @@
+    {
+        patch.BiomeTemplate.Background = Background;
+        patch.CurrentSnapshot.Background = Background;
+        patch.BiomeTemplate.Sunlight.Colour = new Color(0.8f, 0.8f, 1, 1);
+        patch.CurrentSnapshot.LightColour = new Color(0.8f, 0.8f, 1, 1);
+    }
+
+    private void AdjustEnvironment(Patch patch)
+    {
+        var currentTemperature = patch.Biome.ChangeableCompounds[Compound.Temperature];
+        var currentSunlight = patch.Biome.ChangeableCompounds[Compound.Sunlight];
+
+        currentTemperature.Ambient = random.Next(-1, 5) - currentTemperature.Ambient;
+        currentSunlight.Ambient = 0.5f - currentSunlight.Ambient;
+
+        var changes = new Dictionary<Compound, float>
+        {
+            [Compound.Temperature] = currentTemperature.Ambient,
+            [Compound.Sunlight] = currentSunlight.Ambient,
+        };
+
+        patch.Biome.ApplyLongTermCompoundChanges(patch.BiomeTemplate, changes, new Dictionary<Compound, float>());
+    }
+
+    /// <summary>
+    ///   Gets chunks from Iceshelf patch template and applies them to the patches.
+    /// </summary>
+    private void AddIceChunks(Patch patch)
+    {
+        var templateBiome = SimulationParameters.Instance.GetBiome(TemplateBiomeForIceChunks);
+        foreach (var configuration in IceChunksConfigurations)
+        {
+            var iceChunkConfiguration = templateBiome.Conditions.Chunks[configuration];
+            iceChunkConfiguration.Density *= 10;
+            patch.Biome.Chunks.Add(Prefix + configuration, iceChunkConfiguration);
+        }
+    }
+
+    private void LogEvent(Patch patch, double totalTimePassed)
+    {
+        patch.LogEvent(new LocalizedString("GLOBAL_GLACIATION_EVENT"),
+            true, true, "GlobalGlaciationEvent.svg");
+
+        if (patch.Visibility == MapElementVisibility.Shown)
+        {
+            targetWorld.LogEvent(new LocalizedString("GLOBAL_GLACIATION_EVENT_LOG", patch.Name),
+                true, false, "GlobalGlaciationEvent.svg");
+        }
+
+        patch.AddPatchEventRecord(WorldEffectVisuals.GlobalGlaciation, totalTimePassed);
+    }
+
+    private void LogBeginningOfGlaciation()
+    {
+        targetWorld.LogEvent(new LocalizedString("GLOBAL_GLACIATION_START_EVENT_LOG"),
+            true, true, "GlobalGlaciationEvent.svg");
+    }
+
+    private void LogEndOfGlaciation()
+    {
+        targetWorld.LogEvent(new LocalizedString("GLOBAL_GLACIATION_END_EVENT_LOG"),
+            true, true, "GlobalGlaciationEvent.svg");
+    }
+
+    private void FinishEvent()
+    {
+        hasEventAlreadyHappened = true;
+        foreach (var index in modifiedPatchesIds)
+        {
+            if (!targetWorld.Map.Patches.TryGetValue(index, out var patch))
+                continue;
+
+            PatchSnapshot patchSnapshot = patch.History[eventDuration];
+
+            ResetBackground(patch, patchSnapshot);
+            ResetEnvironment(patch, patchSnapshot);
+            RemoveChunks(patch);
```

Comment:
For multiple events to work concurrently, I think we need like a flag in the history snapshot to tell which events are active, that way the history can be scanned for when no event is active. Imagine that we add like a volcano event, and things go like:

- gen 1 no event
- gen 2 volcano event starts
- gen 3 glaciation starts
- gen 4 volcano ends and it reverts things (this can be figured out later but I think the history format should be fixed now)
- gen 6 glaciation ends and it reverts things back to gen 2 *which is incorrect* as the light values will be lower than the default in gen 1

So I think the history needs a flag for active events and probably a helper method for finding conditions without events to revert to by looking backwards.

Reply:
I mostly agree with you but I think that the system for the patch events overlapping should be added and thought about when we are going to add a new patch event. After this event I'll probably start working on another one so it will be much easier and clearer for me to do in the next PR

Reply:
Alright.

I realized after reading my previous message again that my design would be insufficient as we need like a full system that is given the set of active patch events and then determines what patch conditions should be set to. Such a system should be able to correctly handle events starting and stopping at any time and be able to apply any mix of active patch events.

Reply:
This will be handled in the meteor event PR.